compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -779,6 +784,46 @@ fn check_link_name_xor_ordinal(
     }
 }
 
+struct LinterStateMixedExportNameAndNoMangle {
+    export_name: Option<Span>,
+    hir_id: Option<HirId>,
+    no_mangle: Option<Span>,
+}
+
+impl LinterStateMixedExportNameAndNoMangle {
+    fn new() -> Self {
+        Self { export_name: None, hir_id: None, no_mangle: None }
+    }
+
+    fn track_export_name(&mut self, span: Span, hir_id: HirId) {
+        self.export_name = Some(span);
+        self.hir_id = Some(hir_id);
+    }
+
+    fn track_no_mangle(&mut self, span: Span) {
+        self.no_mangle = Some(span.clone());
+    }
+
+    /// Emit diagnostics if the lint condition is met.
+    fn emit_diagnostics_in_case(self, tcx: TyCtxt<'_>) {
+        if let (Some(export_name), Some(no_mangle), Some(hir_id)) =
+            (self.export_name, self.no_mangle, self.hir_id)
+        {
+            tcx.node_span_lint(
+                lint::builtin::MIXED_EXPORT_NAME_AND_NO_MANGLE,
+                hir_id,
+                export_name,
+                |lint| {
+                    lint.primary_message("the attribute `export_name` may not be used in combination with `no_mangle`")
+                        .span_label(export_name, "`export_name` takes precedence")
+                        .span_note(no_mangle.clone(), "the `no_mangle` attribute is ignored")
+                        .span_suggestion_verbose(no_mangle, "remove the `no_mangle` attribtute", "", lint::Applicability::MachineApplicable);
+                },
```

Comment:
As for the `LintDiagnostic` (instead of the manual calls), I think you can add it to the `compiler/rustc_codegen_ssa/src/errors.rs` file of this crate, and use [`emit_node_span_lint`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.emit_node_span_lint) above.

Reply:
Thanks for pointing me in this direction, I added the first `LintDiagnostic` to the error file, see 71e9a91bae4cd140b2dd01adcafefd70c91fff5e